### PR TITLE
Volume snapshot copy

### PR DESCRIPTION
**Pull request by erikh**
This implements `volcli volume snapshot` commands:

* `volcli volume snapshot list` will list all snapshots for a given volume
* `volcli volume snapshot copy` will copy a snapshot from volume A into new volume B (under the same policy)

Please take a close look here as a lot has changed.

/cc @jainvipin @mapuri @unclejack

### CODE REVIEW

**Comment by erikh** [comment]
@mapuri this is ready for review when you have time.

File: storage/backend/ceph/ceph.go [hunk 1]
```diff
@@ -321,6 +320,32 @@ func (c *Driver) ListSnapshots(do storage.DriverOptions) ([]string, error) {
...
+		return err
+	}
+
+	if er.ExitStatus != 0 {
+		return fmt.Errorf("Protecting snapshot for clone (volume %q, snapshot %q): %v", do.Volume.Name, snapName, err)
+	}
+
+	cmd = exec.Command("rbd", "clone", do.Volume.Name, newName, "--snap", snapName, "--pool", do.Volume.Params["pool"])
+	er, err = executor.NewWithTimeout(cmd, do.Timeout).Run()
+	if err != nil {
```
Comment (mapuri) [inline storage/backend/ceph/ceph.go]: I think we should unprotect the sanpshot if clone fails, right?

On a related note, does the snapshot needs to stay protected after clone is successful?

**Comment by mapuri** [comment]
@erikh 

from ux and REST endpoint (`/copy`) perspective I think another alternative would have been to add a flag to the `volume create` sub-command itself, like:

```
volcli volume create --from-snap <> <policy>/<volume>
```

I wondering if the above UX is better as when we integrate this capability in the docker's storage-plugin API then we could just pass `--opt from-snap=<>` in `docker volume create`, which we could then pass to the volplugin's volume create call?

Also I don't see a docker side of integration in this PR, should we add it here in case it helps us make a more informed decision? Or do you plan to add it later?

File: storage/backend/ceph/ceph.go (hunk 1 above)
Comment (erikh) [inline storage/backend/ceph/ceph.go]: correct on the unprotect.

yes, the snapshot cannot be unprotected once the copy is made

**Comment by erikh** [comment]
That's not going to work. The reason is, we need to copy the volume to a new policy before, and the snap id itself is scoped to the existing volume. This won't work.

That said, we could extend it in the future to work this way with a combination of options (for source and snap id). I don't think it's necessary right now, though.

**Comment by mapuri** [comment]
@erikh 

> The reason is, we need to copy the volume to a new policy before, and the snap id itself is scoped to the existing volume

may be i missed this, where does this happens? from my reading of `(d *DaemonConfig) handleCopy()` I was under the impression we create a new volume with exact same policy config as that of the volume to which the snapshot belongs.

so, just to clarify I got it correct, following work flow will work and create `vol2` with old version of policy:

```
volcli upload policy1 < intent-big-size-with-snap
volcli volume create policy1/vol1
volcli upload policy1 < intent-small-size
volcli volume snapshot copy policy1/vol1 snap1 vol2
```

> we could extend it in the future to work this way with a combination of options (for source and snap id). I don't think it's necessary right now, though.

sgtm

**Comment by erikh** [comment]
Correct.

**Comment by mapuri** [comment]
ok, thanks for confirming @erikh . We can revisit UX later when we integrate this feature with docker's API.

LGTM

**Pull request merged**